modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java

```diff
@@ -0,0 +1,1461 @@
+                .collect(Collectors.toSet());
+
+            Set<Integer> retain = new HashSet<>(grps);
+            retain.retainAll(snpGrps);
+
+            if (!retain.isEmpty()) {
+                sctx.acceptException(new IgniteCheckedException("Snapshot has been interrupted due to some of the required " +
+                    "cache groups stopped: " + retain));
+            }
+        }
+    }
+
+    /**
+     * @param snpName Unique snapshot name.
+     * @param srcNodeId Node id which cause snapshot operation.
+     * @param parts Collection of pairs group and appropratate cache partition to be snapshotted.
+     * @param snpSndr Factory which produces snapshot receiver instance.
+     * @return Future which will be completed when snapshot is done.
+     */
+    IgniteInternalFuture<Boolean> runLocalSnapshotTask(
+        String snpName,
+        UUID srcNodeId,
+        Map<Integer, GridIntList> parts,
+        Executor exec,
+        SnapshotFileSender snpSndr
+    ) {
+        if (!busyLock.enterBusy())
+            return new GridFinishedFuture<>(new IgniteCheckedException("Snapshot manager is stopping [locNodeId=" + cctx.localNodeId() + ']'));
+
+        try {
+            SnapshotTask snpTask = putSnapshotTask(snpName,
+                cctx.localNodeId(),
+                parts,
+                snpRunner,
+                snpSndr);
+
+            IgniteInternalFuture<Void> startFut = snpTask.submit(dbMgr::addCheckpointListener,
+                dbMgr::removeCheckpointListener);
+
+            // Snapshot is still in the INIT state. beforeCheckpoint has been skipped
+            // due to checkpoint aready running and we need to schedule the next one
+            // right afther current will be completed.
+            dbMgr.forceCheckpoint(String.format(SNAPSHOT_CP_REASON, snpName));
+
+            startFut.get();
+
+            return snpTask.snapshotFuture();
+        }
+        catch (IgniteCheckedException e) {
+            return new GridFinishedFuture<>(e);
+        }
+        finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /**
+     * @param snpName Unique snapshot name.
+     * @param srcNodeId Node id which cause snapshot operation.
+     * @param parts Collection of pairs group and appropratate cache partition to be snapshotted.
+     * @param snpSndr Factory which produces snapshot receiver instance.
+     * @return Snapshot operation task which should be registered on checkpoint to run.
+     * @throws IgniteCheckedException If fails.
+     */
+    SnapshotTask putSnapshotTask(
+        String snpName,
+        UUID srcNodeId,
+        Map<Integer, GridIntList> parts,
+        Executor exec,
+        SnapshotFileSender snpSndr
+    ) throws IgniteCheckedException {
+        if (locSnpTasks.containsKey(snpName))
+            throw new IgniteCheckedException("Snapshot with requested name is already scheduled: " + snpName);
+
+        SnapshotTask snpTask = locSnpTasks.computeIfAbsent(snpName,
```

Comment:
From concurrency point of view it's better to use here `if (putIfAbsent != null) throw exception`, since there can be race between containsKey and computeIfAbsent.

Reply:
Fixed.